### ISSUE TITLE
JSONPath support

### ISSUE THREAD

**Issue by devAarno** [devAarno/lemon-server#26]
**Story**
I as a developer want to parse some data by a Lemon Server. Let data is a JSON document. I want to get some pieces of a JSON from a request's body.

**Design**
Let size of a JSON document is unlimited in a general case. It means that JSON may be bigger that internal buffers of the Lemon Server. So we have to realize a streaming parser of the body. Lemon Server may read the big body by chuncks but as raw bytes.

The JSON allows us to make requests to its structure by a JSONPath. The JSONPath is convinient and easy to understand expect of its filters like `$..book[?(@.isbn)]` that may be a very complex expression with logic operators.

So, solution consists of:
* a limited parser of JSONPath expressions. Let supports [features](https://goessner.net/articles/JsonPath/) :
  * a root object `$`;
  * a child operator `.` or `[]`;
  * a recursive descent `..`;
  * a wildcard (all objects/elements regardless their names) `*`;
* a full featured JSON parser which supports [ECMA-404](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf);
* a body engine as an user level abstraction. It should hide parsers and the streaming mechanisms;
* a convenient user side. In my vision it would be better to avoid callbacks as a communication between an user code and the parser. As for ANSI C an API based on while-loop is more understandable. So, it should looks like:
```c
static char jsonPath[] = "$.book[2]";
promise p = appendJSONPath(jsonPathCollection, jsonPath);

while (END != (currentFinding = findInJSONBody(body, jsonPathCollection))) {
  switch (currentFinding) {
    case p : puts(p.value);
    default : ...
  }
}
```
Expected that `switch .. case` should be optimised by popular compilers.

**Limitations**
* promises return [`string`](https://github.com/devAarno/lemon-server/blob/b6e40b6a1c3603f0150cb14cad0ccdbb9ca81fc4/src/lemonHttp/string.h#L25) only;
* promises return full values only. If value is bigger than internal buffer, an error will returns. Chunked values unsupported.

**DoD**
* code is written;
* a demo application is preperared;
* test are written.

**Comment by devAarno** [comment]
As I understand, I have to rebuild a JSON parser to support a streaming feature, but I can't to avoid Lemon's restrictions.

**Comment by devAarno** [comment]
As I understand, there is no any good specification of the JSONPath with detailed rules description. I have used the [JSONPath Online Elevator](https://jsonpath.com/) for testing. It seems, there are few extensions implemented here (for example, a recursive ending rule). I found the [JSONPath Plus](https://github.com/JSONPath-Plus/JSONPath) extension. I do not understand its behavior in some tricky cases (for example, [recursive rule with null values](https://github.com/JSONPath-Plus/JSONPath/issues/165)). Anyway, I have spent too much time for JSONPath support implementation. My tests show that it works somehow. The code gets stinky, so I have to refactor it to move on, but the active development block it. So, I consider that JSONPath reaches some milestone and ready to merge in the main branch. This task is going to spit into subtasks further.
